dubbo-registry/dubbo-registry-zookeeper/src/main/java/org/apache/dubbo/registry/zookeeper/ZookeeperRegistry.java

```diff
@@ -126,6 +151,41 @@ public void doUnregister(URL url) {
         }
     }
 
+    private void checkZkProvider(){
+        Set<URL> registeredURLSet = new HashSet<URL>(getRegistered());
+        for (URL registeredURL: registeredURLSet){
+            if(registeredURL.getProtocol().equalsIgnoreCase(Constants.CONSUMER))
+                continue;
+
+            String host = registeredURL.getHost();
+            int port = registeredURL.getPort();
+            String interfaceName = registeredURL.getServiceInterface();
+            List<String> childPath = zkClient.getChildren(Constants.PATH_SEPARATOR + DEFAULT_ROOT + Constants.PATH_SEPARATOR
+                    + interfaceName + Constants.PATH_SEPARATOR + Constants.PROVIDERS_CATEGORY);
+
+            boolean serviceExist = false;
+            for(String child : childPath) {
+                URL providerURL = URL.valueOf(URL.decode(child));
+                if(providerURL != null && providerURL.getHost().equalsIgnoreCase(host) && providerURL.getPort() == port){
+                    serviceExist = true;
+                    break;
+                }
+            }
+
+            if(!serviceExist){
+                try {
+                    logger.error("found some interface" + interfaceName + " does not exists in zk provider path, do recover");
+                    recover();
+                } catch (Exception t) {
+                    logger.error("Unexpected error occur at failed checkZk, cause: " + t.getMessage(), t);
+                }
+                return;

```

Comment:
Sorry for asking this question (as I am new to dubbo at this part). Is it expected that if a single service (from group of services) does not exists then we will stop the recovery for remaining services, as I see "return" statement in the code?